geopandas/sindex.py

```diff
@@ -1,25 +1,341 @@
+            geo_index = []
+
+            for i, geo in enumerate(geometry):
+                res = self.query(geo, predicate=predicate, sort=sort)
+                if res.size > 0:
+                    # sort results and append
+                    tree_index.extend(res)
+                    geo_index.extend([i] * len(res))
+            return np.vstack([geo_index, tree_index])
+
+        @property
+        def size(self):
+            return len(self.leaves()[0][1])
+
+        @property
+        def is_empty(self):
+            return self.size == 0
+
+
+if compat.HAS_PYGEOS:
+
+    from pygeos import STRtree, box, points, Geometry  # noqa
+
+    class PyGEOSSTRTreeIndex(STRtree):
+        """
+        A simple wrapper around pygeos's STRTree
+        """
+
+        with_objects = namedtuple("with_objects", "object id")
+
+        def __init__(self, geometry):
+            # for compatibility with old RTree implementation, store ids/indexes
+            original_indexes = geometry.index
+            non_empty = geometry[~geometry.values.is_empty]
+            self.objects = self.ids = original_indexes[~geometry.values.is_empty]
+            super().__init__(non_empty.values.data)
+
+        def query_bulk(self, geometry, predicate=None, sort=False):
+            """Wrapper to expose underlaying pygeos objects to pygeos.query_bulk.
+            This also allows a deterministic (sorted) order for the results.
+
+            Parameters
+            ----------
+            geometry : {GeoSeries, GeometryArray, numpy.array of PyGEOS geometries}
+                Accepts GeoPandas geometry iterables (GeoSeries, GeometryArray)
+                or a numpy array of PyGEOS geometries.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                The predicate to use for testing geometries from the tree
+                that are within the input geometry's envelope.
+                See `pygeos.strtree` documentation for more details on
+                predicates.
+            sort : bool, default False
+                Determines if the results should be sorted lexicographically using
+                geometry's indexes as the primary key and the sindex's indexes as the
+                secondary key.
+
+            Returns
+            -------
+            ndarray with shape (2, n)
+                The first subarray contains input geometry indexes.
+                The second subarray contains tree geometry indexes.
+
+            See also
+            --------
+            See PyGEOS.strtree documentation for more information.
+            """  # noqa: E501
+            try:
+                # for GeoSeries or GeometryArray
+                geometry = geometry.geometry.values.data
+            except AttributeError:
+                # assume already a numpy array of pygeos geoms
+                pass
+            res = super().query_bulk(geometry, predicate)
```

Comment:
Similar to comment above, I don't think the raw results can be returned when inputs have an underlying index.  `query_bulk` returns the index (position) within the ndarray of values, not the index values from the series index.  Same for the indexes into the tree geometries.

Personally I'm also not keen on use of `try` / `AttributeError`; I think more explicit handling of supported types would be better.

Untested code as an example:
```
index = None
if isinstance(geometry, GeoSeries):
    geometry = geometry.values.data
    index = geometry.index
elif isinstance(geometry, GeometryArray):
    geometry = geometry.data
elif not isinstance(geometry, np.ndarray):
    raise ValueError(...)

res = super().query_bulk(geometry, predicate)

# convert back to original indices
if index is not None:
    res[0] = index[res[0]]

res[1] = self.ids[res[1]]
```

Reply:
Yeah, what I have is kind of ugly, I like what you have better, I'll try to implement it.